csharp/autobuilder/Semmle.Autobuild.CSharp/DotNetRule.cs

```diff
@@ -70,9 +70,18 @@ public BuildScript Analyse(IAutobuilder<CSharpAutobuildOptions> builder, bool au
                 });
         }
 
-        private static BuildScript WithDotNet(IAutobuilder<AutobuildOptionsShared> builder, Func<string?, IDictionary<string, string>?, BuildScript> f)
+        /// <summary>
+        /// Returns a script that attempts to download relevant version(s) of the
+        /// .NET Core SDK, followed by running the script generated by <paramref name="f"/>.
+        ///
+        /// The arguments to <paramref name="f"/> are the path to the directory in which the
+        /// .NET Core SDK(s) were installed and any additional required environment
+        /// variables needed by the installed .NET Core (<code>null</code> when no variables
+        /// are needed).
+        /// </summary>
+        public static BuildScript WithDotNet(IAutobuilder<AutobuildOptionsShared> builder, Func<string?, IDictionary<string, string>?, BuildScript> f)
         {
-            var installDir = builder.Actions.PathCombine(builder.Options.RootDirectory, ".dotnet");
+            var installDir = builder.Actions.PathCombine(Path.GetTempPath(), "GitHub", ".dotnet");
             var installScript = DownloadDotNet(builder, installDir);
```

Comment:
Why not use `$CODEQL_EXTRACTOR_CSHARP_SCRATCH_DIR`, when it is set? That dir is database-creation specific, and will be cleaned up after extraction.

Reply:
I think there's a chance that `$CODEQL_EXTRACTOR_CSHARP_SCRATCH_DIR` might be inside the repo folder.

Reply:
Right, if the user creates the DB inside the repo. But is that problematic?

Reply:
We would be extracting more DLLs in that case, because we extract all DLLs inside the repo folder. There's no point in extracting asp.net core DLLs or SDK specific DLLs, such as csc.dll.

Reply:
We should probably just let that logic skip the scratch dir?

Reply:
I have reverted my commit which installs .NET to a temporary directory. Instead, we now use `git ls-files "*.dll"` to find DLL files (unless path(s) are specified with `--references`) so the problem of having extra DLLs somewhere in the workspace goes away.